### PR TITLE
Adding more logging and adding consistent patterns

### PR DESCRIPTION
**Pull request by rturner3**
Signed-off-by: Ryan Turner <turner@uber.com>

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**
- Error logging in public Handler methods
- Error wrapping and logging in private methods and functions

**Description of change**
- Adding common handler for logging and wrapping errors. The general pattern being followed is that downstream errors should be logged as an "error" field, and then the additional error message string at the current layer should be logged as the message. The return error value should look like "newErr[: oldErr]" to preserve the stack trace.
  Since the stack trace is high cardinality, only log the message generated from the current method context and keep all other error details in the "error" field.
- Removing high cardinality values from log message strings. Prefer to log these as fields to simplify log querying.
- Fixing unit tests which were verifying log messages and error strings that conflicted with the previously described error handling and logging pattern.

**Which issue this PR fixes**
N/A

### CODE REVIEW

**Comment by evan2645** [comment]
Thanks for your work here @rturner3.

Sorry if this sounds a bit aloof, but would you mind quickly and succinctly describing what the goal of this change is? As far as I can tell, it is related to beefing up logging in spire-server handler related code, but also something about doing that consistently? I have a few thoughts on the change but am curious to hear more about the intention first

**Comment by rturner3** [comment]
Sure, no problem, @evan2645.

TL;DR
My overall intent here is to improve observability into when server RPC calls are failing. I noticed a few improvement opportunities in this area:
- There are quite a few places where server RPC methods are returning errors, but those errors are not being logged.
- Errors that are the return value of calling some other method are being logged with the full stack of errors in the message, which results in a higher cardinality log message set.
- High cardinality parameters are being included in some error strings and are being logged as part of the message.

**Longer explanation**
Most of these changes are focused on logging errors whenever public Handler methods implementing a gRPC interface method return some error. Another aspect of these changes was to enact some consistency in how these errors from Handler methods are logged.

For public Handler methods implementing a gRPC interface method, I tried following these patterns:
- Log the error returned from the public method, which includes the gRPC status, as the log message i.e. the result of `status.Error()`
- If the error being logged from the public method was due to some downstream error from another method, log the downstream error in the "error" field of the log. i.e.
```
err := doSomething()
if err {
    topLevelErr := status.Error(codes.InvalidArgument, "some high level error message")
    log.WithField(telemetry.Error, err).Error(topLevelErr)
    return topLevelErr
}
```

While making these changes, I noticed some long error messages where the downstream error was a stack of nested errors, e.g. `top-level error: second-level error: third-level error`. In order to avoid these kind of cases where error messages logged from private methods become high cardinality strings that are more taxing to index, I also introduced an `errorutil` package for wrapping and logging errors. The utility methods in that package generate error logs with the "error" field set to `second-level error: third-level error` and the error message as `top-level error`. Then the downstream error (which may or may not be a stack of errors) is wrapped into a new error with the full stack of errors and returned to the caller, e.g. `top-level error: second-level error: third-level error`.

There were also a few cases where high cardinality values were being included as part of the error message itself using `fmt.Errorf()`. I changed these occurrences to use log fields to capture the high cardinality values and omit them from the error string so that they are not included in any error string stack passed back to the caller. e.g.
Instead of:
```
err = fmt.Errorf("invalid parameter foo %v", foo)
log.Error(err)
return err // if caller logs this returned err, they get the high cardinality value in the message
```
prefer this pattern:
```
err = errors.New("invalid parameter foo")
log.WithField(telemetry.Foo, foo).Error(err)
return err
```

**Comment by azdagron** [comment]
Huge 💯 to logging the errors before returning a failure to an RCP client. Thanks @rturner3! I did notice a few spots where it looks like We'll end up logging strings starting with a lower-case letter. We've settled on errors starting with lowercase and logging always starting with upper case. Can you fix those spots?

**Comment by rturner3** [comment]
> Huge 💯 to logging the errors before returning a failure to an RCP client. Thanks @rturner3! I did notice a few spots where it looks like We'll end up logging strings starting with a lower-case letter. We've settled on errors starting with lowercase and logging always starting with upper case. Can you fix those spots?

Sure, just pushed a commit which attempts to address that convention as well. Let me know if you have any questions or concerns on the approach I took.

**Comment by rturner3** [comment]
Ok, I have this back into a reviewable state now. Thanks for all the constructive feedback so far!

File: pkg/agent/endpoints/workload/handler.go [hunk 1]
```diff
@@ -145,22 +145,27 @@ func (h *Handler) FetchJWTBundles(req *workload.JWTBundlesRequest, stream worklo
 
 // ValidateJWTSVID processes request for JWT SVID validation
 func (h *Handler) ValidateJWTSVID(ctx context.Context, req *workload.ValidateJWTSVIDRequest) (*workload.ValidateJWTSVIDResponse, error) {
+	log := h.Log.WithField(telemetry.Method, telemetry.ValidateJWTSVID)
 	if req.Audience == "" {
+		log.Error("audience must be specified")
```
Comment (azdagron) [inline pkg/agent/endpoints/workload/handler.go]: Logs should start with capitalization but audience refers to an argument, so it would be weird. Maybe

```
log.Errorf("Missing required audience parameter")
```

File: pkg/agent/endpoints/workload/handler.go [hunk 2]
```diff
@@ -145,22 +145,27 @@ func (h *Handler) FetchJWTBundles(req *workload.JWTBundlesRequest, stream worklo
...
 		return nil, status.Error(codes.InvalidArgument, "audience must be specified")
 	}
+
+	log = log.WithField(telemetry.Audience, req.Audience)
 	if req.Svid == "" {
+		log.Error("svid must be specified")
```
Comment (azdagron) [inline pkg/agent/endpoints/workload/handler.go]: Same thing here...

File: pkg/agent/endpoints/workload/handler.go [hunk 3]
```diff
@@ -145,22 +145,27 @@ func (h *Handler) FetchJWTBundles(req *workload.JWTBundlesRequest, stream worklo
...
 		return nil, status.Error(codes.InvalidArgument, "svid must be specified")
 	}
 
 	_, selectors, metrics, done, err := h.startCall(ctx)
 	if err != nil {
+		log.WithError(err).Error("Failed to validate JWT SVID during context parsing")
```
Comment (azdagron) [inline pkg/agent/endpoints/workload/handler.go]: we should be consistent in regards to `JWT SVID` v.s. `JWTSVID`  (and `JWT-SVID` for that matter)

File: pkg/agent/endpoints/workload/handler.go [hunk 4]
```diff
@@ -231,23 +237,21 @@ func (h *Handler) sendX509SVIDResponse(update *cache.WorkloadUpdate, stream work
...
-		log.Error(err)
-		return err
+		log.WithError(err).Error("Could not serialize X.509 SVID response")
+		return status.Errorf(codes.Unavailable, "could not serialize response: %v", err)
 	}
 
 	err = stream.Send(resp)
 	if err != nil {
-		log.Error(err)
+		log.WithError(err).Error("Stream send of X.509 SVID response failed")
```
Comment (azdagron) [inline pkg/agent/endpoints/workload/handler.go]: nit: "Failed to send X.509 SVID response"

File: pkg/agent/endpoints/workload/handler.go (hunk 3 above)
Comment (rturner3) [reply pkg/agent/endpoints/workload/handler.go]: Good point. I see that the terminology is currently inconsistent across the code comments, log messages, and documentation as is on master. Just confirming, are you asking me to change only the instances I'm touching in this PR, or in all places?

It looks like we should probably stick to `JWT-SVID` when referring to the identity document construct and `JWTSVID` when referring to the specific types defined in the node API proto and SPIRE codebase.

Comment (azdagron) [reply pkg/agent/endpoints/workload/handler.go]: The workload handler at a minimum should probably be consistent. I haven't looked though to know if it was already consistent or not. Your calculus for deciding which term to use sounds good.

Comment (rturner3) [reply pkg/agent/endpoints/workload/handler.go]: Updated instances in the code I touched and other places in the handler files I touched to be consistent in using `JWT-SVID` over `JWT SVID` to align with the terminology defined in the SPIFFE specification. If we want to clean up other usages of `JWT SVID`, we can probably create a new issue for that work since this PR is already pretty broad in scope.
